### PR TITLE
PauseActivity implementation

### PR DESCRIPTION
**Pull request by ychebotarev**
## What changed?
Add PuaseActivity API implementation

* Add Paused field to the pending activity info
* Add replication for that field
* Add populating search attributes
* basic testing

## Why?
Working on activity API

## How did you test it?
Add basic unit tests. Func tests will come later.

### CODE REVIEW

File: proto/internal/temporal/server/api/historyservice/v1/request_response.proto
Comment (bergundy) [inline proto/internal/temporal/server/api/historyservice/v1/request_response.proto:617]: How about we store `pause_info` with a struct instead of a boolean? It'll let us surface more information at a glance, such as who initiated pausing and when.

File: service/history/api/pauseactivity/api.go
Comment (bergundy) [inline service/history/api/pauseactivity/api.go:49]: Might as well use `GetWorkflowId()` and `GetRunId()` to avoid panics in case `FrontendRequest` is nil.

File: service/history/api/pauseactivity/api.go
Comment (bergundy) [inline service/history/api/pauseactivity/api.go:75]: This isn't related to this PR, but can we call this API `PauseActivityById` (and name the request and response accordingly)? It matches the frontend API and leaves room for an API that works on an activity by token or event ID if we ever consider adding one.

File: service/history/api/pauseactivity/api.go
Comment (bergundy) [inline service/history/api/pauseactivity/api.go:88]: Hmm... should we store a request ID and error if the pause request came with a different request ID?

File: service/history/api/pauseactivity/api.go
Comment (bergundy) [inline service/history/api/pauseactivity/api.go:96]: This should be renamed to match the new search attribute name.

File: service/history/api/pauseactivity/api_test.go
Comment (bergundy) [inline service/history/api/pauseactivity/api_test.go:142]: check that the activity was actually paused?

Comment (bergundy) [reply service/history/api/pauseactivity/api_test.go:142]: Also not great practice to test unexported functions. TBH, I'd just do a functional test here, this test doesn't provide much value.

File: service/history/api/respondactivitytaskfailed/api_test.go
Comment (bergundy) [inline service/history/api/respondactivitytaskfailed/api_test.go:541]: Hmm... why are we not recording the last started time anymore?

File: service/history/api/pauseactivity/api.go
Comment (bergundy) [inline service/history/api/pauseactivity/api.go:75]: We could also just move all of this into MutableStateImpl to avoid scattering the activity logic further in the code base.

File: service/history/timer_queue_standby_task_executor.go
Comment (bergundy) [inline service/history/timer_queue_standby_task_executor.go:328]: Wouldn't the stamp be updated in the standby cluster if the activity is paused? Why not use the same mechanism?

File: service/history/workflow/mutable_state.go
Comment (bergundy) [inline service/history/workflow/mutable_state.go:367]: Do you really need this method?
I saw that you use it in the api implementation, but you already have the `*persistencespb.ActivityInfo`, so why not pass that into the already existing `UpdateActivity`?

File: service/history/workflow/mutable_state_impl.go
Comment (bergundy) [inline service/history/workflow/mutable_state_impl.go:5000]: I think there's a lot more code organization that we could do here. I wouldn't call this specific TODO here but 🤷.

File: service/history/workflow/mutable_state_impl.go
Comment (bergundy) [inline service/history/workflow/mutable_state_impl.go:4998]: Should this be called after we update any activity so we don't forget to call it when necessary or are we concerned with the potential overhead?

File: service/history/workflow/mutable_state_impl.go
Comment (bergundy) [inline service/history/workflow/mutable_state_impl.go:5012]: You don't need to escape if the set of value types is limited and cannot contain colons. Escaping was only needed when we considered packing three values into this attribute.

File: service/history/workflow/mutable_state_impl.go
Comment (bergundy) [inline service/history/workflow/mutable_state_impl.go:5013]: I would denote this is an entity type: `entityType:activity:%s`

File: service/history/workflow/mutable_state_impl.go
Comment (bergundy) [inline service/history/workflow/mutable_state_impl.go:5016]: This isn't a policy, this is a manual pause request, e.g: `reason:manual:%s` and let the user specify the reason so this can be used to find all affected workflows.

Comment (bergundy) [reply service/history/workflow/mutable_state_impl.go:5016]: I would put this in the loop that goes over all of the paused activities so you can make a more local decision on the value here while maintaining the set and dedupe semantics.

File: service/history/workflow/mutable_state_impl.go
Comment (bergundy) [inline service/history/workflow/mutable_state_impl.go:4998]: This should be called after applying replication tasks too.

**Review by bergundy — COMMENTED** [review]
Logic seems good to me. The stamp approach is really paying off.

File: service/history/api/pauseactivity/api.go
Comment (ychebotarev) [reply service/history/api/pauseactivity/api.go:49]: Request doesn't have "GetWorkflowId()" or "GetRunId()"

File: service/history/api/pauseactivity/api.go
Comment (ychebotarev) [reply service/history/api/pauseactivity/api.go:75]: In the beginning we have both functions - with  ByID and without it(ByToken). We remove  "ByToken" version after very long discussion.

File: service/history/api/pauseactivity/api.go
Comment (ychebotarev) [reply service/history/api/pauseactivity/api.go:88]: why that should be an error? Pausing paused activity just do nothing, but didn't fail.

File: service/history/workflow/mutable_state_impl.go
Comment (ychebotarev) [reply service/history/workflow/mutable_state_impl.go:4998]: this I didn't understand. Why? This will add _another_ search attribute? Or we don't replicate visibility?

File: service/history/api/pauseactivity/api.go
Comment (ychebotarev) [reply service/history/api/pauseactivity/api.go:96]: done

File: service/history/api/respondactivitytaskfailed/api_test.go
Comment (ychebotarev) [reply service/history/api/respondactivitytaskfailed/api_test.go:541]: RecordLastActivityStarted is renamed to RecordLastActivityCompleteTime, because this is what it actually doing (LastAttemptCompleteTime).

File: service/history/api/pauseactivity/api.go
Comment (ychebotarev) [reply service/history/api/pauseactivity/api.go:75]: I would rather put it into activity.go. I don't want to keep it in mutable state - it is already overcrowded. 
Also in the future we can move it into state machine.

File: service/history/workflow/mutable_state_impl.go
Comment (ychebotarev) [reply service/history/workflow/mutable_state_impl.go:5012]: In our documentation activity type is defined as "is the mapping of a name to an Activity Definition." There is no restrictions. In probably most SDK activity type is just a function name, which of cause don't need to be escaped. But we don't guarantee that, at least I don't see it in the documentation

File: service/history/workflow/mutable_state_impl.go
Comment (ychebotarev) [reply service/history/workflow/mutable_state_impl.go:5013]: what is the value of adding 'entityType'?

File: service/history/api/pauseactivity/api.go
Comment (bergundy) [reply service/history/api/pauseactivity/api.go:49]: But `FrontendRequest` does.

File: service/history/api/pauseactivity/api.go
Comment (bergundy) [reply service/history/api/pauseactivity/api.go:75]: So should we rename the history service API to be more specific?

File: service/history/api/pauseactivity/api.go
Comment (bergundy) [reply service/history/api/pauseactivity/api.go:88]: To give an indication to the user that the activity was already paused and their request didn't take effect.

File: service/history/workflow/mutable_state_impl.go
Comment (bergundy) [reply service/history/workflow/mutable_state_impl.go:5013]: To give the full namespace for the entity type and remove the possibility of future collisions.

File: service/history/workflow/mutable_state_impl.go
Comment (bergundy) [reply service/history/workflow/mutable_state_impl.go:4998]: We don't replicate visibility tasks.

**Comment by bergundy** [comment]
This is also missing functional and XDC tests. Can come in a followup PR.

File: service/history/api/respondactivitytaskfailed/api_test.go
Comment (bergundy) [reply service/history/api/respondactivitytaskfailed/api_test.go:541]: I'm probably missing some context here but we also sometimes record the start time in mutable state, where does this happen now (to save me from reading)?

File: service/history/workflow/mutable_state.go
Comment (bergundy) [reply service/history/workflow/mutable_state.go:367]: Why is that a concern?

File: service/history/workflow/mutable_state_impl.go
Comment (ychebotarev) [reply service/history/workflow/mutable_state_impl.go:5013]: then it should be "entityType.activity"

File: proto/internal/temporal/server/api/historyservice/v1/request_response.proto
Comment (ychebotarev) [reply proto/internal/temporal/server/api/historyservice/v1/request_response.proto:617]: no at this iteration (will be in batch activity API)

File: service/history/api/pauseactivity/api.go
Comment (ychebotarev) [reply service/history/api/pauseactivity/api.go:49]: ah, ok I didn't get it

File: service/history/api/pauseactivity/api_test.go
Comment (ychebotarev) [reply service/history/api/pauseactivity/api_test.go:142]: Everything should be unit-tested is possible.
We can't test that activity was actually paused, mock mutable state is used.
And yes, there will be functional tests on top of that, just not in this PR.

File: service/history/api/pauseactivity/api.go
Comment (ychebotarev) [reply service/history/api/pauseactivity/api.go:75]: ok

File: service/history/timer_queue_standby_task_executor.go
Comment (ychebotarev) [reply service/history/timer_queue_standby_task_executor.go:328]: .Puased check is not needed

File: service/history/workflow/mutable_state.go
Comment (ychebotarev) [reply service/history/workflow/mutable_state.go:367]: note: one more refactoring PR to remove old UpdateActivity

File: service/history/workflow/mutable_state_impl.go
Comment (ychebotarev) [reply service/history/workflow/mutable_state_impl.go:5012]: ok lets remove it

File: service/history/workflow/mutable_state_impl.go
Comment (ychebotarev) [reply service/history/workflow/mutable_state_impl.go:5016]: lets wait

Comment (ychebotarev) [reply service/history/workflow/mutable_state_impl.go:5016]: remove until we better understand what we want

File: service/history/api/pauseactivity/api.go
Comment (ychebotarev) [reply service/history/api/pauseactivity/api.go:49]: done

File: service/history/api/pauseactivity/api.go
Comment (ychebotarev) [reply service/history/api/pauseactivity/api.go:75]: renamed, moved PauseActivityById to activity.go

File: service/history/workflow/mutable_state.go
Comment (ychebotarev) [reply service/history/workflow/mutable_state.go:367]: https://temporalio.atlassian.net/browse/OSS-3579

File: service/history/workflow/mutable_state_impl.go
Comment (ychebotarev) [reply service/history/workflow/mutable_state_impl.go:5000]: ok removed

File: service/history/workflow/mutable_state_impl.go
Comment (ychebotarev) [reply service/history/workflow/mutable_state_impl.go:4998]: refactored.
This call will happen in UpdateActivityCallback, if Paused property changed.

File: service/history/workflow/mutable_state_impl.go
Comment (ychebotarev) [reply service/history/workflow/mutable_state_impl.go:5013]: renamed

File: service/history/workflow/mutable_state_impl.go
Comment (ychebotarev) [reply service/history/workflow/mutable_state_impl.go:4998]: ok, then it needs to be re-applied.
I update UpdateActivityInfo (it is called after replication to sync activities)

File: service/history/workflow/activity.go
Comment (bergundy) [inline service/history/workflow/activity.go:185]: Still think there's value in using request ID for idempotency and otherwise returning an error.
Worth bringing up for discussion if you disagree.
Not blocking this PR.

File: service/history/workflow/mutable_state.go
Comment (bergundy) [inline service/history/workflow/mutable_state.go:367]: Does this need to be exported?

File: service/history/workflow/mutable_state.go
Comment (bergundy) [inline service/history/workflow/mutable_state.go:368]: Don't export IMO, this can be an implementation detail.

File: service/history/workflow/mutable_state_impl.go
Comment (bergundy) [inline service/history/workflow/mutable_state_impl.go:4995]:
```suggestion
	// if prev found it can point to the same activity info as incoming activity info.
```

File: service/history/workflow/mutable_state_impl.go
Comment (bergundy) [inline service/history/workflow/mutable_state_impl.go:5030]: nit: I still think using `:` as the sole namespace delimiter is the right way to go here.

File: service/matching/matcher_test.go
Comment (bergundy) [inline service/matching/matcher_test.go:425]:
```suggestion
```

**Review by bergundy — APPROVED** [review]
Approved with comments, feel free to merge after you've address the ones I've not marked as nits or optional.

File: service/history/workflow/mutable_state_impl.go
Comment (alexshtin) [inline service/history/workflow/mutable_state_impl.go:5030]: Why do you need `entityType` here?

**Review by alexshtin — APPROVED** [review]

**Review by alexshtin — APPROVED** [review]

File: service/history/workflow/activity.go
Comment (ychebotarev) [reply service/history/workflow/activity.go:185]: I do disagree, I think we should raise it with product. Usually it is better to NOT return an error - users can add some alerts/metrics to errors. Anyway, lets raise it with Drew (though - when?)

File: service/history/workflow/mutable_state.go
Comment (ychebotarev) [reply service/history/workflow/mutable_state.go:367]: yes, it is used outside on mutable_state_impl.go

File: service/history/workflow/mutable_state.go
Comment (ychebotarev) [reply service/history/workflow/mutable_state.go:368]: yes, with the new changes this one shouldn't be expoerted.

File: service/history/workflow/mutable_state_impl.go
Comment (ychebotarev) [reply service/history/workflow/mutable_state_impl.go:4995]: done

File: service/history/workflow/mutable_state_impl.go
Comment (ychebotarev) [reply service/history/workflow/mutable_state_impl.go:5030]: will discuss separately.

File: service/matching/matcher_test.go
Comment (ychebotarev) [reply service/matching/matcher_test.go:425]: hm, interesting. this was after merge.

File: service/history/api/pauseactivity/api.go
Comment (ychebotarev) [reply service/history/api/pauseactivity/api.go:56]: done

File: service/history/workflow/mutable_state_impl.go
Comment (ychebotarev) [reply service/history/workflow/mutable_state_impl.go:4991]: may be, but after refactoring (remove UpdateActivity)

File: service/history/workflow/mutable_state_impl.go
Comment (ychebotarev) [reply service/history/workflow/mutable_state_impl.go:5026]: will discuss offline

File: service/history/workflow/mutable_state_impl.go
Comment (ychebotarev) [reply service/history/workflow/mutable_state_impl.go:4986]: I'm going to remove `UpdateActivity` method and replace it with this one.
The whole idea is - we capture the size of "prev" in mutable state, then call update callback.

Comment (ychebotarev) [reply service/history/workflow/mutable_state_impl.go:4986]: actually I like the idea

File: service/history/workflow/mutable_state_impl.go
Comment (ychebotarev) [reply service/history/workflow/mutable_state_impl.go:5026]: discussed.
agree on
```
“property:activityType=MYBESTACTIVITYTYPE”

“property:workflowType=MYWORKFLOWID”
```

File: service/history/timer_queue_active_task_executor.go
Comment (ychebotarev) [reply service/history/timer_queue_active_task_executor.go:520]: ok, but I think it is somewhat missleading to return NotFound if it was paused.

Comment (ychebotarev) [reply service/history/timer_queue_active_task_executor.go]: done

File: service/history/workflow/mutable_state_impl.go
Comment (ychebotarev) [reply service/history/workflow/mutable_state_impl.go]: done

File: service/history/workflow/timer_sequence.go
Comment (ychebotarev) [reply service/history/workflow/timer_sequence.go:209]: I think about it - no, I don't need it. 

GenerateActivityRetryTasks is protected
I add a Paused check if task refresher.
As for the GenerateActivityTasks - in all other cases activity can't be paused.

Note: this will change as we move to paused rules.

File: service/history/workflow/mutable_state_impl.go
Comment (xwduan) [reply service/history/workflow/mutable_state_impl.go:4992]: This should be user data change and should be added into the new map?

Comment (ychebotarev) [reply service/history/workflow/mutable_state_impl.go:4992]: I have no idea what it means

Comment (xwduan) [reply service/history/workflow/mutable_state_impl.go:4992]: Sorry, I was replying to @yycptt . The context is in this pr: https://github.com/temporalio/temporal/pull/6816

**Pull request merged**
